jetty-server/src/main/java/org/eclipse/jetty/server/Request.java

```diff
@@ -367,6 +368,39 @@ public void addEventListener(final EventListener listener)
         if (listener instanceof AsyncListener)
             throw new IllegalArgumentException(listener.getClass().toString());
     }
+    
+    /**
+     * Remember a session that this request has just entered.
+     * 
+     * @param s the session
+     */
+    public void enterSession(final HttpSession s)
+    {
+        if (s == null)
+            return;
+
+        if (_sessions == null)
+            _sessions = new ArrayList<>();
+        if (LOG.isDebugEnabled())
+            LOG.debug("Request {} entering session={}", this, s);
+        _sessions.add(s);
+    }
+
+    /**
+     * Complete this request's access to a session.
+     *
+     * @param s the session
+     */
+    private void leaveSession(final HttpSession s)
+    {
+        if (s == null)
+            return;
+        
+        Session session = Session.class.cast(s);
+        if (LOG.isDebugEnabled())
+            LOG.debug("Request {} leaving session {}", this, session);
+        session.getSessionHandler().complete(session);
```

Comment:
This strikes me as asymmetric: why the "enter" event does not need to tell the `SessionHandler` but adds the session to the list, but the "leave" event needs to tell to the `SessionHandler` and _not_ remove the session from the list?
I mean I can see the list is cleared out in `recycle()`, but often asymmetries of this kind are indication of a problem.

Reply:
The SessionHandler tells the Request that it has entered the Session. The Request knows when it is completing, and tells the SessionHandler that it is leaving the Session. The Request could iterate over the list and remove each Session as it calls its SessionHandler with it. I just didn't bother because the next thing that happens is that the Request gets recycled.